modularity: depsolving (COMPOSER-2365)

supakeen opened the pull request:
The commit implements modularity for the dnf4 dependency solver.

The dependency solver now takes two new keys per transaction: `module-install-specs` and `module-enable-specs`, each of which is a list of module specifications.

When modules are used and packages for them are installed a new dictionary is returned in the response where each key is the name of an enabled module and the value are the contents of the module configuration and the module failsafe file that DNF needs in the resulting system.

These values can be used by consumers of the dependency solver, such as `images`, to create the appropriate inline sources and copy stages.

---

When `dnf` enables a module on a system it writes a file in `/etc/dnf/modules.d/{name}.module` with the following contents:

```
[nodejs]
name=nodejs
stream=18
profiles=common
state=enabled
```

it also drops a file in `/var/lib/dnf/modulefailsafe/` containing the module metadata yaml so it has access to it if the repository that the module belongs to gets disabled or removed.

This approach seems to be the most lightweight way to deal with modularity even if it *feels* hacky. The alternatives involve the complex task(s) of exporting and importing dnf state.

This is a draft PR to discuss the approach, I'm assuming naming and implementation details can be figured out when we undraft.

My big questions are:

1. Do we want to instead make our own format that the solver returns and then have specific stages that understand that format to write the module and failsafe file?
2. Nonexisting modules seem to not fail the depsolve.
3. Do we want *both* `install` *and* `enable`?
4. The interaction with our idea of transactions and enabled/installed modules is finicky. As long as we do all the modular stuff in a single transaction (the second one, when coming from `images`) all should be peachy but there are pitfalls if we were to add a third transaction after this.

---

If/after we decide this PR is the correct approach this will need tests.

If/after we merge this PR we will need a PR to `images` to create the appropriate copy stages and for our customizations to pass in modular names.

For proper testing we'll also need to include module metadata in `rpmrepo` and the mock repositories we have in here.

---

If you want to test this you can check out this branch and use the following:

```
user@muja osbuild € cat request.json 
{
  "command": "depsolve",
  "arch": "x86_64",
  "module_platform_id": "platform:c9s",
  "releasever": "9",
  "cachedir": "/tmp",
  "arguments": {
    "root_dir": "/tmp",
    "repos": [
      {
        "id": "AppStream",
        "metalink": "https://mirrors.centos.org/metalink?repo=centos-appstream-9-stream&arch=$basearch&protocol=https,http"
      },
      {
        "id": "BaseOS",
        "metalink": "https://mirrors.centos.org/metalink?repo=centos-baseos-9-stream&arch=$basearch&protocol=https,http"
      }
    ],
    "transactions": [
      {
        "module-install-specs": ["nodejs:18/minimal"],
        "package-specs": ["@core"],
        "exclude-specs": []
      }
    ]
  }
}
user@muja osbuild € ./tools/osbuild-depsolve-dnf < request.json | jq .modules
{
  "nodejs": {
    "module-file": "[nodejs]\nname=nodejs\nstream=18\nprofiles=common\nstate=enabled\n",
    "failsafe-file": "---\ndocument: modulemd\nversion: 2\ndata:\n  name: nodejs\n  stream: \"18\"\n  version: 920240311154052\n  context: rhel9\n  static_context: true\n  arch: x86_64\n  summary: Javascript runtime\n  description: >-\n    Node.js is a platform built on Chrome's JavaScript runtime for easily building\n    fast, scalable network applications. Node.js uses an event-driven, non-blocking\n    I/O model that makes it lightweight and efficient, perfect for data-intensive\n    real-time applications that run across distributed devices.\n  license:\n    module:\n    - MIT\n    content:\n    - MIT\n    - MIT and ASL 2.0 and ISC and BSD\n  xmd: {}\n  dependencies:\n  - buildrequires:\n      platform: [el9]\n  references:\n    community: http://nodejs.org\n    documentation: http://nodejs.org/en/docs\n    tracker: https://github.com/nodejs/node/issues\n  profiles:\n    common:\n      rpms:\n      - nodejs\n      - npm\n    development:\n      rpms:\n      - nodejs\n      - nodejs-devel\n      - npm\n    minimal:\n      rpms:\n      - nodejs\n    s2i:\n      rpms:\n      - nodejs\n      - nodejs-nodemon\n      - npm\n  api:\n    rpms:\n    - nodejs\n    - nodejs-devel\n    - npm\n  buildopts:\n    rpms:\n      macros: >\n        %_with_bundled 1\n\n        %_with_python3_fixup 0\n  components:\n    rpms:\n      nodejs:\n        rationale: Javascript runtime and npm package manager.\n        ref: stream-nodejs-18-rhel-9.4.0\n        buildorder: 100\n        arches: [aarch64, ppc64le, s390x, x86_64]\n      nodejs-nodemon:\n        rationale: Simple monitor script for use during development of a node.js app\n        ref: stream-nodejs-18-rhel-9.4.0\n        buildorder: 200\n        arches: [aarch64, ppc64le, s390x, x86_64]\n      nodejs-packaging:\n        rationale: RPM Macros and Utilities for Node.js Packaging\n        ref: stream-nodejs-18-rhel-9.4.0\n        buildorder: 10\n        arches: [aarch64, ppc64le, s390x, x86_64]\n  artifacts:\n    rpms:\n    - nodejs-1:18.19.1-1.module_el9+986+cd5c1200.src\n    - nodejs-1:18.19.1-1.module_el9+986+cd5c1200.x86_64\n    - nodejs-debuginfo-1:18.19.1-1.module_el9+986+cd5c1200.x86_64\n    - nodejs-debugsource-1:18.19.1-1.module_el9+986+cd5c1200.x86_64\n    - nodejs-devel-1:18.19.1-1.module_el9+986+cd5c1200.x86_64\n    - nodejs-docs-1:18.19.1-1.module_el9+986+cd5c1200.noarch\n    - nodejs-full-i18n-1:18.19.1-1.module_el9+986+cd5c1200.x86_64\n    - nodejs-nodemon-0:3.0.1-1.module_el9+663+27147c5a.noarch\n    - nodejs-nodemon-0:3.0.1-1.module_el9+663+27147c5a.src\n    - nodejs-packaging-0:2021.06-4.module_el9+244+c48c75d4.noarch\n    - nodejs-packaging-0:2021.06-4.module_el9+244+c48c75d4.src\n    - nodejs-packaging-bundler-0:2021.06-4.module_el9+244+c48c75d4.noarch\n    - npm-1:10.2.4-1.18.19.1.1.module_el9+986+cd5c1200.x86_64\n...\n"
  }
}
```

---

- [ ] The selected profile is hardcoded, this needs to be parsed from the module specification *or* parsed from the module metadata (default profile) if we can't get it from `dnf`.

croissanne commented on osbuild/solver/dnf.py:
```diff
@@ -322,9 +337,78 @@ def depsolve(self, arguments):
+                    [{module.getName()}]
+                    name={module.getName()}
+                    stream={module.getStream()}
+                    profiles=common
```
What if a module doesn't have a common profile? Like the postgresql module only has

``` 
...
  profiles:
    client:
      rpms:
      - postgresql
    server:
      rpms:
      - postgresql-server
...
```

Should we just add all profiles?

supakeen replied:
Ah, good catch yes. We either parse the module specification and see if it contains a selected profile or we parse the module metadata and select the default profile (as `dnf` does).

We can't really select all profiles as that would lead to a mismatch of what `dnf` installed and what it thinks is installed.

croissanne commented:
So in general I think this is the easiest way forward: drop-in file + the failsafe.

Regarding the profiles, we could either decide to make users choose, or do something silly like heuristically decide which profile by querying what profiles a ModulePackage is part of, and then pick whichever profile isn't common between all of them.

That said, it might be more straight-forward to just let users select the profile, so I'm leaning towards that atm.

croissanne commented:
Having thought about it a bit more, I think we should just expose the having to choose the profiles for now. So you'd choose a module stream + profile.

supakeen commented:
> Having thought about it a bit more, I think we should just expose the profiles for now. So you'd choose a modulestream + profile. And then the module_install makes sense, so we can drop the module_enable.

So, does this mean that we'll *require* a profile to be given (e.g. we'll always require a full module specification for the depsolver)? Seems to be the most proper to me. Then higher abstraction levels can handle which one they select.

croissanne commented:
> > Having thought about it a bit more, I think we should just expose the profiles for now. So you'd choose a modulestream + profile. And then the module_install makes sense, so we can drop the module_enable.
> 
> So, does this mean that we'll _require_ a profile to be given (e.g. we'll always require a full module specification for the depsolver)? Seems to be the most proper to me. Then higher abstraction levels can handle which one they select.

Yup!

thozza reviewed (commented):
This looks amazingly simple, well done! :clap: 

I'm not a modules expert, but my understanding was that one could already depsolve modules with our implementation, and the problem was that we would throw away any module-related files created as a result of DNF (auto-enabling a module). It is desirable to enable a specific module stream (not the default) because there may not be any default, or the default may already be EOL.

If this is true, my naive understanding is that this PR could be simplified to:

- Add an option to enable a specific module stream (`module-enable-specs`)
- Determine if any of the `package-specs` that start with `@` correspond to modules, using `base_module.get_modules()`.
- Determine if any packages in the resulting transaction match packages in modules obtained by `base_module.get_modules()` and append the necessary module-related information to the depsolve result.

Do you think that it would make sense to extend the `rpm` stage to accept the module-related sources as input and handle the correct setup before installing any RPMs? I know that the `copy` stage could do it, but in this case, I'd favor keeping this related functionality contained in a single stage.

thozza commented:
> My big questions are:
> 
> 1. Do we want to instead make our own format that the solver returns and then have specific stages that understand that format to write the module and failsafe file?

I don't see a big use case for this unless we'd like to support enabling specific modules (streams) without installing any packages from it. I guess that we won't need this.

> 2. Nonexisting modules seem to not fail the depsolve.

It would be great if they would :thinking: 

> 3. Do we want _both_ `install` _and_ `enable`?

I'm pretty sure that we need to be able to enable non-default module streams. As for the installation, I'd like us to handle streams within `package_specs` if possible.

> 4. The interaction with our idea of transactions and enabled/installed modules is finicky. As long as we do all the modular stuff in a single transaction (the second one, when coming from `images`) all should be peachy but there are pitfalls if we were to add a third transaction after this.

Did you already run into any specific issues while experimenting with this? It would be great to have a test for such a scenario, but at the end of the day, we should be fine with just two transactions.

supakeen commented:
> > My big questions are:
> > 
> > 1. Do we want to instead make our own format that the solver returns and then have specific stages that understand that format to write the module and failsafe file?
> 
> I don't see a big use case for this unless we'd like to support enabling specific modules (streams) without installing any packages from it. I guess that we won't need this.
> 
> > 2. Nonexisting modules seem to not fail the depsolve.
> 
> It would be great if they would 🤔

I'll see what I can do today :)

> > 3. Do we want _both_ `install` _and_ `enable`?
> 
> I'm pretty sure that we need to be able to enable non-default module streams. As for the installation, I'd like us to handle streams within `package_specs` if possible.

Right, so when a module is installed it means 'install all packages this (default) profile in this module'. When a module is enabled it will then be used during the normal package install.

> > 4. The interaction with our idea of transactions and enabled/installed modules is finicky. As long as we do all the modular stuff in a single transaction (the second one, when coming from `images`) all should be peachy but there are pitfalls if we were to add a third transaction after this.
> 
> Did you already run into any specific issues while experimenting with this? It would be great to have a test for such a scenario, but at the end of the day, we should be fine with just two transactions.

No, we almost always have two transactions; things would only get weird if we had modules enabled in the first *and* second.

supakeen commented:
PR updated; it can now be reviewed (tests pending). I've removed the `installs-specs` after a chat with @thozza. Did this in a separate commit so it can be reverted.

It turns out that requesting a non-existent module *does* throw a marking error so we're also good on that front.

croissanne commented:
> > My big questions are:
> > 
> > 1. Do we want to instead make our own format that the solver returns and then have specific stages that understand that format to write the module and failsafe file?
> 
> I don't see a big use case for this unless we'd like to support enabling specific modules (streams) without installing any packages from it. I guess that we won't need this.
> 
> > 2. Nonexisting modules seem to not fail the depsolve.
> 
> It would be great if they would 🤔
> 
> > 3. Do we want _both_ `install` _and_ `enable`?
> 
> I'm pretty sure that we need to be able to enable non-default module streams. As for the installation, I'd like us to handle streams within `package_specs` if possible.

Are you sure? I think you can just do `$module:$stream/$profile`? And from the UI we would always pass along the stream version and the profile. So I think we should just be able to call `module_install`?

schuellerf commented on osbuild/solver/dnf.py:
```diff
@@ -322,9 +337,78 @@ def depsolve(self, arguments):
+                    transaction.get("module-install-specs", []),
+                    transaction.get("module-enable-specs", []),
+                ):
+                    # we don't particularly care about the NSVCAP here, just the module
```
Name, Stream, Version, Context, Architecture, Profile ?

thozza commented:
> Are you sure? I think you can just do `$module:$stream/$profile`? And from the UI we would always pass along the stream version and the profile. So I think we should just be able to call `module_install`?
> 
> edit: I think module_install should work ,let's just do that and just require module, stream and profile from the user.

[`install_specs` calls `module_install` for modules.](https://github.com/rpm-software-management/dnf/blob/067591db69d6fb5be22856cdb85beb8d7032b86d/dnf/base.py#L2039-L2042)

Looking at the DNF code, we could probably just pass `$module:$stream/$profile` to `install_specs`.

The question is how end users tend to install packages from modules. With your suggested approach, users won't be able to selectively install specific packages from a specific `$module:$stream`. Enabling a specific module/stream makes its content available for installation AFAICT. Without enabling modules, they will be able to only install specific module/stream/profile.

supakeen commented:
Thank you for the expanded reply. I feel like these decisions are best left to the frontend where users get to select modules and packages. I've reverted the commit here that allows to only install modules in favor of having both `enable-modules` and `install-modules`.

I've also rewritten the logic that deals with the profiles that end up in the profile configuration file, it now correctly handles cases where no profiles where requested, or multiple.

supakeen commented:
I don't understand what `autopep8` is trying to do here but I don't want it?

croissanne commented on osbuild/solver/dnf.py:
```diff
@@ -261,7 +268,14 @@ def depsolve(self, arguments):
+                # module
+                self.base_module.enable(transaction.get("module-enable-specs", []))
+
+                # installing a module takes the specification of the module and then
+                # installs all packages belonging to it
+                self.base_module.install(transaction.get("module-install-specs", []))
```
I'm not sure if this is still valid, but tbh it might be nice to have it as an explicit spec, for amongst other things the logic down below?

supakeen replied:
I kept in (both) approaches on purpose to make sure that we don't block a frontend from choosing their preferred option later (likely; only enable).

croissanne reviewed (dismissed):
ty!

thozza commented on osbuild/solver/dnf.py:
```diff
@@ -322,9 +338,83 @@ def depsolve(self, arguments):
+                # packages-to-install set so let's do this only once here
+                package_nevras = list(p["nevra"] for p in packages)
+
+                for module_spec in itertools.chain(
+                    transaction.get("module-install-specs", []),
+                    transaction.get("module-enable-specs", []),
+                ):
+                    module_packages, module_nsvcap = self.base_module.get_modules(module_spec)
```
Here, I'd iterate over `module-enable-specs` and `install-specs` (I think modules must be prefixed with `@`) and call `self.base_module.get_modules()` on them to determine if they correspond to a module.

supakeen replied:
I've applied this change.

thozza reviewed (changes requested):
I like this PR, but I'm not on board with `module-install-specs`. I still think we should stick only to `install-specs`, let the front end do whatever it needs to, and add modules to the list with any other packages or groups to be installed in the image.

In addition, I'd like to see the new functionality to be covered by tests before merging.

croissanne reviewed (approved):
thanks!

achilleas-k commented on osbuild/solver/dnf.py:
```diff
@@ -322,9 +338,88 @@ def depsolve(self, arguments):
+        for module_ns, (module, profiles) in modules.items():
+            profiles = ",".join(profiles) if profiles else ""
+
+            response["modules"][module.getName()] = {
+                "module-file": {
+                    "data": textwrap.dedent(f"""\
+                        [{module.getName()}]
+                        name={module.getName()}
+                        stream={module.getStream()}
+                        profiles={profiles}
+                        state=enabled
+                        """),
+                    "path": f"/etc/dnf/modules.d/{module.getName()}.conf",
+                },
+                "failsafe-file": {
+                    "data": module.getYaml(),
+                    "path": f"/var/lib/dnf/modulefailsafe/{module.getName()}:{module.getStream()}",
+                },
+            }
```
I'm not a fan of this.  Why return a text blob in a structured response?  The way it's done here is understandably straightforward for the use case of generating the inline config file, but I also think moving the inline bits around isn't a great idea either.

The depsolve flow right now doesn't assume much about how the response will be used.  Consumers of the depsolver can use it to generate sources and stage options, which is the primary use case, but you can also use it to generate a package listing for metadata or a report of some kind.  Making the depsolver return text blobs limits the usefulness.

I think it's much more preferable to return structured data and let consumers deal with transforming it as they need.  The transformation is trivial anyway, just like it is with the package metadata when they get turned into sources and stage options.

And speaking of stage options: I'm not a fan of inlining well structured config files that we want to support widely.  We've been doing this more and more lately and I don't like it.  It feels like we're abandoning the idea of osbuild stages with structured options and we'll soon end up with whole chunks of pipelines that are just copy stages from inline sources.

supakeen replied:
Yes, this is the `1.` question in the original PR. We spoke about it though it seems only in non-public places and previously we felt that something that'll only be supported on RHEL 8 and RHEL 9 (maybe) wasn't enough reason to expend the extra effort of making structured data and new stages that would then immediately take that data and string-format in the structured data.

I agree with the concern, it's however a bunch of extra work for in this case very little benefit (nothing else will use these stages), if other things would use these stages it'd tip it the other way for me.

croissanne replied:
How about we make the module-file structured, with a specific stage option. And the failsafe-file, which is really just a blob, a copy from inline source / no separate stage needed?

supakeen commented:
I have added the c9s appstream metadata to our test repositories; I have added a test case that installs a module (and verifies that this is returned).

It doesn't quite validate the entire return value but I think this is enough of a start to get this in.

mvo5 commented on test/data/testrepos/appstream/repodata/repomd.xml:
```diff
@@ -0,0 +1,80 @@
+    <open-size>51840420</open-size>
+  </data>
+  <data type="filelists">
+    <checksum type="sha256">72d4b6933f70291d407371a5ebc4cdb26b3e0ea37b176657a8f3e4d40d9f32f6</checksum>
```
Silly question, could we just remove the filelist from the repomd.xml and also from this PR somehow? Or play some tricks to achieve that?  It makes me sad that we need to add 17Mb of binary data to our git repo that seems unnecessary for the test.

supakeen replied:
I took this as they were also present in the other repositories; they're being tested in the combinations for metadata I believe. cc @thozza.

thozza replied:
Disclaimer: I don't know the answer.

From what I learned about repository metadata, the `filelists` contain information about the content (files and directories) of the packages in the repository. So, not including it may, in theory, work for basic depsolving. However, we can't remove it from all repositories, because the information about provided files and dependencies on them is used to determine dependencies between packages when generating SBOMs. This is also done in our tests.

So removing `filelists` may be OK, but it may also impact SBOM testing. However, it is theoretically possible to delete them from the newly added appstream repo (I think), but no guarantees.

mvo5 replied:
I would really like us to explore how to avoid embedding 18mb compressed binary data into our git tree for this new test. For filelists.xml.gz that is already in the tree it is already in the tree, nothing can be done. Maybe there is a stable URL to download it from (a lot of our tests require downloading/caching stuff and that would be an incrediable cheap/simple option)? I can also explore if pruning it/removing it is possible, a quick way to find out would be to just delete it, remove it from the repomd.xml and see if this one new particular test still generates the expected results. I can look into this next week if required.

supakeen replied:
The filelists are explicitly tested; if you want to fiddle with removing them or not you can do so. This is now the only thing blocking this PR.

mvo5 replied:
I looked at this now (sorry that it took me so long) and I think https://github.com/osbuild/osbuild/commit/b1a4e287146d6c82293d17ee8683f5f9424a7fe2 is a way out and avoids the 18mb by just hand-crafting an empty filelist. 

Please have a look It will need to be squashed into the commit that adds the filelist (with an explanation about the "why") so that we never have the 18mb in the git history. I did not test with dnf5, it may need tweaks there (dnf4 is pretty liberal with the checking, if e.g. open checksum does not match it does not care on my dnf4 version).

thozza replied:
My only worry is that this may affect the SBOM tests, but from what I can tell, we are not explicitly testing dependencies on files provided by a package, so this approach will probably be acceptable. Specifically, this piece of the code https://github.com/osbuild/osbuild/blob/1a529d78a97c6f661f3ecad46b75a696ae0ed379/osbuild/util/sbom/dnf.py#L101-L103

mvo5 replied:
Note that my patch only removed the appstream filelist, we can add a comment in the test that appstream packaes have an empty filelist. But this should not affect anything existing and we can keep the existing filelists as removing them does not really help anything (they are already part of the git history)

thozza replied:
Don't get me wrong, I'm all for removing it if it does not break existing tests...

supakeen commented:
This PR has been updated with a stage to write DNF's module configuration files with. This stage is *very wide* as I couldn't quite grok how tight we want to make the pattern value regular expressions.

thozza commented:
Sorry, I accidentally pushed a commit to this branch, which was intended for a different PR. I fixed it now. :innocent:

thozza commented:
> @thozza Could you help me figure out the depsolving testing failures here? After adding the AppStream repositories DNF5 is now returning different package sets from those returned by DNF4. Perhaps they're not receiving the same repositories?

I've started with the low-hanging fruit - the `dump` test case. This one needs fixing the expected package list OR (IMHO, a better solution because of "4444" vs. "21577" packages with the appstream repo) not use the appstream repository for this test case. The command dumps the content of all configured repositories, so by adding a new one, you'll get packages from it as well.

I'm still looking at the depsolve test case...

thozza commented:
For transparency, I've run into broken dependencies issues while using the `baseos` and `appstream` repos in combination with DNF, because these are not based on the same snapshot. So, both repos should be based on the same snapshot and updated in lockstep.

supakeen commented:
I've updated the BaseOS snapshots (in the same commit as the appstream adding) and updated the package sets to conform to the new snapshots (`initscripts-service`) was dropped.

thozza commented:
Unfortunately, it seems that DNF5 behavior is different than DNF4, when installing the `@core` group to an empty root :facepalm: 

I've run:
```console
$ rpm -q python3-dnf
python3-dnf-4.22.0-1.fc41.noarch

$ sudo dnf4 --assumeno --nogpgcheck --installroot /tmp/dnf-installroot/ \
    --repofrompath='baseos,file:///home/thozza/devel/osbuild/test/data/testrepos/baseos' \
    --repofrompath='custom,file:///home/thozza/devel/osbuild/test/data/testrepos/custom' \
    --repofrompath='appstream,file:///home/thozza/devel/osbuild/test/data/testrepos/appstream' \
    --repo baseos --repo appstream --repo custom \
    install '@core' > dnf4-installroot-core.log
```
[dnf4-installroot-core.log](https://github.com/user-attachments/files/18451610/dnf4-installroot-core.log)

```console
$ rpm -q dnf5
dnf5-5.2.8.1-3.fc41.x86_64

$ sudo dnf5 --assumeno --nogpgcheck --installroot /tmp/dnf-installroot/ \
    --repofrompath='baseos,file:///home/thozza/devel/osbuild/test/data/testrepos/baseos' \
    --repofrompath='custom,file:///home/thozza/devel/osbuild/test/data/testrepos/custom' \
    --repofrompath='appstream,file:///home/thozza/devel/osbuild/test/data/testrepos/appstream' \
    --repo baseos --repo appstream --repo custom \
    install '@core' > dnf5-installroot-core.log
```
[dnf5-installroot-core.log](https://github.com/user-attachments/files/18451612/dnf5-installroot-core.log)

The side-by-side diff:
[dnf-installroot-core-diff.txt](https://github.com/user-attachments/files/18451618/dnf-installroot-core-diff.txt)

It turns out that DNF4 would install **323** packages, while DNF5 would install **318** packages... :fire: 

My suggestion would be to add the `appstream` repository only for the modules test case and leave it out in the other test cases...

thozza commented:
I've created https://github.com/rpm-software-management/dnf5/issues/2013

supakeen commented:
I've added a new `enabled_repos` key to the testcases; only the repositories listed in this key are passed to each respective testcase. I've then reset all testcases to only use `baseos` and `custom` (as it was before); and only enable appstream for the `basic_module` depsolve testcase.

croissanne reviewed (dismissed):
lgtm! ty

achilleas-k reviewed (approved):
Thanks.  LGTM.

Small question in-line.

supakeen commented on stages/org.osbuild.dnf.module-config:
```diff
@@ -0,0 +1,29 @@
+
+    # we need to handle enabled profiles as a string as configparser
+    # would normally write [a, b]
+    inip[conf["name"]]["profiles"] = ", ".join(conf["profiles"])
```
It's indeed not important.

supakeen commented on stages/org.osbuild.dnf.module-config.meta.json:
```diff
@@ -0,0 +1,41 @@
+      "properties": {
+        "path": {
+          "type": "string",
+          "description": "Path to write the module configuration to."
```
I've dropped the config name/path from the stage entirely; instead opting to deduce it from the module name.

supakeen commented on osbuild/solver/dnf.py:
```diff
@@ -322,9 +337,85 @@ def depsolve(self, arguments):
+        modules = {}
+
+        for transaction in transactions:
+            # module specifications must start with an "@" and include a ":", filter them
+            # out so we can use them
+            modules_in_package_specs = [
+                p[1:] for p in transaction.get("package-specs", []) if p.startswith("@") and ":" in p]
```
I can't. It's probably correct that only using `@` and `base_module.get_modules()` is more appropriate. I'll change that.

I had based this on trying to install modules without specifying streams but it's likely that these modules didn't *have* a default stream :)

supakeen replied:
I've added a commit that relaxes the requirement to only testing for `@`-at-the-start and calls the `get_modules` to ask DNF if something is a module maybe.

thozza reviewed (approved):
Thanks, LGTM :+1:

mvo5 reviewed (changes requested):
I set this to change requested (sorry) because of https://github.com/osbuild/osbuild/pull/1933#discussion_r1924959064 and to avoid that it accidentally gets merged.

supakeen commented:
I have force pushed as I have applied @mvo5's patch to the filelists which makes them empty and that had to be done in the commit that introduced the file(s). Sorry for dismissing all reviews :(
